### PR TITLE
PooledByteBufferAllocator port, Byte Buffers improvements

### PR DESCRIPTION
**Pull request by nayato**
Motivation:
Existing rudimentary buffer pool implementation is not flexible to efficiently support scenarios with high variety in message sizes.

Modifications:
- implemented PooledByteBufferAllocator (port of PooledByteBufAllocator) and related artifacts
- made IByteBuffer implement IEquatable<IByteBuffer>, IComparable<IByteBuffer>, implemented Equals, GetHashCode.
- ported test suite for Byte Buffers and fixed discovered issues.
- added suppression of ExecutionContext flow on async read / write on socket-based channels.
- removed old pooled Byte Buffer allocator implementation.

Result:
Support for buffer pooling for larger byte sequences, more efficient usage of memory dedicated for the pool.

### CODE REVIEW

**Comment by nayato** [comment]
@mtuchkov, @mastermanu (@manusr?), please take a look.

File: src/DotNetty.Buffers/PoolArena.cs [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        void AllocateNormal(PooledByteBuffer<T> buf, int reqCapacity, int normCapacity)
+        {
+            ++this.allocationsNormal;
+            if (this.q050.Allocate(buf, reqCapacity, normCapacity) || this.q025.Allocate(buf, reqCapacity, normCapacity)
+                || this.q000.Allocate(buf, reqCapacity, normCapacity) || this.qInit.Allocate(buf, reqCapacity, normCapacity)
+                || this.q075.Allocate(buf, reqCapacity, normCapacity) || this.q100.Allocate(buf, reqCapacity, normCapacity))
```
Comment (myagley) [inline src/DotNetty.Buffers/PoolArena.cs]: The netty version of this function doesn't appear to allocate using the `q100` chunk list. I'm not sure if that is a bug in netty or an oversight here.

Comment (nayato) [inline src/DotNetty.Buffers/PoolArena.cs]: recent change: https://github.com/netty/netty/commit/1b4a5609d31e667a268fd882883a3b6ff356dfb8
I'll go through to see if I'm missing smth else as well.

**Comment by myagley** [comment]
lgtm

**Comment by myagley** [comment]
except for the merge conflicts and the check failures :wink:

**Comment by nayato** [comment]
`LeakDetectionTest.UnderReleaseBufferLeak` still fails on CI for an unknown reason (but not when run in VS). Has nothing to do with these changes so merging.

**Pull request merged**